File: mqtt_subscriber.py
import json
import os
import re
import ssl
from datetime import datetime
from typing import Callable, Dict, List, Optional, Tuple

import mysql.connector
import paho.mqtt.client as mqtt
from dotenv import load_dotenv

load_dotenv()

MQTT_BROKER = os.getenv('MQTT_BROKER', 'broker.hivemq.com')
MQTT_PORT = int(os.getenv('MQTT_PORT', '8883'))
MQTT_TOPIC = os.getenv('MQTT_TOPIC', '')
MQTT_TOPICS = os.getenv('MQTT_TOPICS', '')
MQTT_USERNAME = os.getenv('MQTT_USERNAME')
MQTT_PASSWORD = os.getenv('MQTT_PASSWORD')
MQTT_TLS_ENABLED = os.getenv('MQTT_TLS_ENABLED', 'true').lower() in ('1', 'true', 'yes')
MQTT_TLS_CAFILE = os.getenv('MQTT_TLS_CAFILE')
MQTT_CLIENT_ID = os.getenv('MQTT_CLIENT_ID', 'viafacil_subscriber')

DB_CONFIG = {
    'host': os.getenv('DB_HOST', 'localhost'),
    'user': os.getenv('DB_USER', 'root'),
    'password': os.getenv('DB_PASSWORD', ''),
    'database': os.getenv('DB_NAME', 'sa_viafacil_db')
}

NUMERIC_REGEX = re.compile(r'-?\d+(?:[\.,]\d+)?')

def parse_numeric_payload(payload: str) -> Optional[float]:
    match = NUMERIC_REGEX.search(payload)
    if not match:
        return None
    value = match.group(0).replace(',', '.')
    try:
        return float(value)
    except ValueError:
        return None

def parse_presence_payload(payload: str) -> Optional[float]:
    normalized = payload.strip().lower()
    if normalized in {'1', 'true', 'ativo', 'on', 'detectado'}:
        return 1.0
    if normalized in {'0', 'false', 'inativo', 'off', 'livre'}:
        return 0.0
    return parse_numeric_payload(payload)

SENSOR_TOPIC_MAP: Dict[str, Dict[str, object]] = {
    'S1 umidade':      {'tipo': 's1_umidade',      'unidade': '%',    'parser': parse_numeric_payload},
    'S1 temperatura':  {'tipo': 's1_temperatura',  'unidade': '°C',   'parser': parse_numeric_payload},
    'S1 iluminacao':   {'tipo': 's1_iluminacao',   'unidade': 'lux',  'parser': parse_numeric_payload},
    'Projeto S2 Distancia1': {'tipo': 's2_distancia1', 'unidade': 'cm',   'parser': parse_numeric_payload},
    'Projeto S2 Distancia2': {'tipo': 's2_distancia2', 'unidade': 'cm',   'parser': parse_numeric_payload},
    'Projeto S3 Presenca3':  {'tipo': 's3_presenca',   'unidade': 'status', 'parser': parse_presence_payload},
    'Projeto S3 Ultrassom3': {'tipo': 's3_ultrassom',  'unidade': 'cm',   'parser': parse_numeric_payload},
    'projeto trem velocidade': {'tipo': 'trem_velocidade', 'unidade': 'km/h', 'parser': parse_numeric_payload},
}

def resolve_topics() -> List[str]:
    explicit = [t.strip() for t in MQTT_TOPICS.replace('"', '').split(',') if t.strip()]
    fallback = MQTT_TOPIC.strip()
    topics = explicit
    if fallback:
        topics.append(fallback)
    return sorted(set(topics))

SUBSCRIPTION_LIST = resolve_topics()

def extract_measurement(topic: str, payload_raw: bytes) -> Tuple[str, float, str]:
    payload = payload_raw.decode(errors='ignore').strip()

    try:
        data = json.loads(payload)
        tipo = data.get('tipo')
        valor = data.get('valor')
        unidade = data.get('unidade')
        if tipo and valor is not None:
            return tipo, float(valor), unidade or ''
    except json.JSONDecodeError:
        pass

    config = SENSOR_TOPIC_MAP.get(topic)
    if not config:
        raise ValueError(f"Tópico '{topic}' não está mapeado e payload não é JSON")

    parser = config.get('parser', parse_numeric_payload)
    valor = parser(payload)
    if valor is None:
        raise ValueError(f"Não foi possível interpretar valor numérico para o tópico '{topic}'")

    return config['tipo'], float(valor), config.get('unidade', '')

def get_db_connection():
    return mysql.connector.connect(**DB_CONFIG)

def get_or_create_sensor(cursor, tipo):
    cursor.execute("SELECT id FROM sensor WHERE tipo = %s AND status = 'ativo' LIMIT 1", (tipo,))
    result = cursor.fetchone()
    
    if result:
        return result[0]
    
    print(f"  [AVISO] Sensor tipo '{tipo}' não encontrado - criando novo sensor...")
    descricao = f"Sensor {tipo} - Auto-criado via MQTT"
    cursor.execute("INSERT INTO sensor (tipo, descricao, status) VALUES (%s, %s, 'ativo')", (tipo, descricao))
    sensor_id = cursor.lastrowid
    print(f"  [OK] Novo sensor criado com ID: {sensor_id}")
    return sensor_id

def on_connect(client, userdata, flags, rc):
    timestamp = datetime.now().strftime('%Y-%m-%d %H:%M:%S')
    if rc == 0:
        print(f"[{timestamp}] ✓ Conectado ao broker com sucesso!")
        topics = SUBSCRIPTION_LIST or [MQTT_TOPIC]
        for topic in topics:
            client.subscribe(topic)
            print(f"[{timestamp}] ✓ Inscrito no tópico: {topic}")
        print(f"[{timestamp}] Aguardando mensagens... (Ctrl+C para sair)")
        print("-" * 60)
    else:
        print(f"[ERRO] Falha na conexão. Código: {rc}")

def on_message(client, userdata, msg):
    timestamp = datetime.now().strftime('%Y-%m-%d %H:%M:%S')
    print(f"\n[{timestamp}] Nova mensagem recebida!")
    print(f"  Tópico: {msg.topic}")
    print(f"  Payload: {msg.payload.decode(errors='ignore')}")

    try:
        tipo, valor, unidade = extract_measurement(msg.topic, msg.payload)
    except ValueError as err:
        print(f"  [ERRO] {err}")
        return
     
    try:
        conn = get_db_connection()
        cursor = conn.cursor()
        
        id_sensor = get_or_create_sensor(cursor, tipo)
        
        cursor.execute(
            "INSERT INTO sensor_data (id_sensor, valor, unidade, data_hora) VALUES (%s, %s, %s, NOW())",
            (id_sensor, valor, unidade)
        )
        
        conn.commit()
        
        print(f"  [✓] Dados salvos no banco com sucesso!")
        print(f"      ID Sensor: {id_sensor} | Valor: {valor} {unidade}")
        
        cursor.close()
        conn.close()
        
    except mysql.connector.Error as err:
        print(f"  [ERRO] Falha no banco de dados: {err}")
    except Exception as e:
        print(f"  [ERRO] Erro inesperado: {e}")

def on_disconnect(client, userdata, rc):
    timestamp = datetime.now().strftime('%Y-%m-%d %H:%M:%S')
    if rc != 0:
        print(f"\n[{timestamp}] [AVISO] Desconectado inesperadamente. Tentando reconectar...")

def main():
    print(f"[{datetime.now().strftime('%Y-%m-%d %H:%M:%S')}] Iniciando cliente MQTT assinante...")
    print(f"[{datetime.now().strftime('%Y-%m-%d %H:%M:%S')}] Conectando ao broker: {MQTT_BROKER}:{MQTT_PORT}")
    if SUBSCRIPTION_LIST:
        print(f"[{datetime.now().strftime('%Y-%m-%d %H:%M:%S')}] Tópicos: {', '.join(SUBSCRIPTION_LIST)}")
    elif MQTT_TOPIC:
        print(f"[{datetime.now().strftime('%Y-%m-%d %H:%M:%S')}] Tópico: {MQTT_TOPIC}")
    print("-" * 60)
    
    client = mqtt.Client(client_id=MQTT_CLIENT_ID, callback_api_version=mqtt.CallbackAPIVersion.VERSION1)

    if MQTT_USERNAME and MQTT_PASSWORD:
        client.username_pw_set(MQTT_USERNAME, MQTT_PASSWORD)

    if MQTT_TLS_ENABLED:
        context = ssl.create_default_context()
        if MQTT_TLS_CAFILE:
            context.load_verify_locations(MQTT_TLS_CAFILE)
        client.tls_set_context(context)
        client.tls_insecure_set(False)
    
    client.on_connect = on_connect
    client.on_message = on_message
    client.on_disconnect = on_disconnect
    client.reconnect_delay_set(min_delay=2, max_delay=30)
    
    try:
        client.connect(MQTT_BROKER, MQTT_PORT, 60)
        client.loop_forever()
    except KeyboardInterrupt:
        print("\n\n[INFO] Encerrando cliente MQTT...")
        client.disconnect()
        print("[INFO] Cliente desconectado com sucesso!")
    except Exception as e:
        print(f"[ERRO] Erro fatal: {e}")

if __name__ == "__main__":
    main()

File: mqtt_subscriber.php
<?php

require_once __DIR__ . '/includes/phpMQTT.php';
require_once __DIR__ . '/config/db.php';

set_time_limit(0);

function loadEnvFile(string $path): void
{
    if (!file_exists($path)) {
        return;
    }

    $lines = file($path, FILE_IGNORE_NEW_LINES | FILE_SKIP_EMPTY_LINES);
    foreach ($lines as $line) {
        $trimmed = trim($line);
        if ($trimmed === '' || $trimmed[0] === '#') {
            continue;
        }

        $parts = explode('=', $trimmed, 2);
        $key = trim($parts[0]);
        $value = trim($parts[1] ?? '');
        $value = trim($value, "\"' ");

        putenv("{$key}={$value}");
        $_ENV[$key] = $value;
    }
}

function env(string $key, $default = null)
{
    $value = getenv($key);
    return ($value === false || $value === null) ? $default : $value;
}

loadEnvFile(__DIR__ . '/.env');

$mqttConfig = [
    'broker' => env('MQTT_BROKER', 'broker.hivemq.com'),
    'port' => (int)env('MQTT_PORT', 8883),
    'username' => env('MQTT_USERNAME', ''),
    'password' => env('MQTT_PASSWORD', ''),
    'tls_enabled' => filter_var(env('MQTT_TLS_ENABLED', 'true'), FILTER_VALIDATE_BOOLEAN),
    'tls_cafile' => env('MQTT_TLS_CAFILE') ?: __DIR__ . '/config/certs/DigiCertGlobalRootCA.crt.pem',
    'client_id' => env('MQTT_CLIENT_ID', 'viafacil_subscriber') . '_' . uniqid(),
    'wildcard_topic' => env('MQTT_TOPIC', 'viafacil/sensores/#'),
    'topics_list' => array_filter(array_map('trim', explode(',', str_replace('"', '', env('MQTT_TOPICS', '')))))
];

$sensorTopicMap = [
    'S1 umidade' => ['tipo' => 's1_umidade', 'unidade' => '%', 'parser' => 'parseNumericPayload'],
    'S1 temperatura' => ['tipo' => 's1_temperatura', 'unidade' => '°C', 'parser' => 'parseNumericPayload'],
    'S1 iluminacao' => ['tipo' => 's1_iluminacao', 'unidade' => 'lux', 'parser' => 'parseNumericPayload'],
    'Projeto S2 Distancia1' => ['tipo' => 's2_distancia1', 'unidade' => 'cm', 'parser' => 'parseNumericPayload'],
    'Projeto S2 Distancia2' => ['tipo' => 's2_distancia2', 'unidade' => 'cm', 'parser' => 'parseNumericPayload'],
    'Projeto S3 Presenca3' => ['tipo' => 's3_presenca', 'unidade' => 'status', 'parser' => 'parsePresencePayload'],
    'Projeto S3 Ultrassom3' => ['tipo' => 's3_ultrassom', 'unidade' => 'cm', 'parser' => 'parseNumericPayload'],
    'projeto trem velocidade' => ['tipo' => 'trem_velocidade', 'unidade' => 'km/h', 'parser' => 'parseNumericPayload'],
];

function parseNumericPayload(string $payload): ?float
{
    if (preg_match('/-?\d+(?:[\.,]\d+)?/', $payload, $matches)) {
        $value = str_replace(',', '.', $matches[0]);
        return is_numeric($value) ? (float)$value : null;
    }
    return null;
}

function parsePresencePayload(string $payload): ?float
{
    $normalized = strtolower(trim($payload));
    return match ($normalized) {
        '1', 'true', 'on', 'ativo', 'detectado' => 1.0,
        '0', 'false', 'off', 'inativo', 'livre' => 0.0,
        default => parseNumericPayload($payload)
    };
}

function extractMeasurement(string $topic, string $payload, array $map): array
{
    $decoded = json_decode($payload, true);
    if (is_array($decoded) && isset($decoded['tipo'], $decoded['valor'])) {
        return [
            'tipo' => (string)$decoded['tipo'],
            'valor' => (float)$decoded['valor'],
            'unidade' => $decoded['unidade'] ?? ''
        ];
    }

    if (!isset($map[$topic])) {
        throw new InvalidArgumentException("Tópico '{$topic}' não está mapeado e payload não é JSON válido");
    }

    $config = $map[$topic];
    $parser = $config['parser'];
    $valor = $parser($payload);

    if ($valor === null) {
        throw new InvalidArgumentException("Não foi possível interpretar valor numérico para o tópico '{$topic}'");
    }

    return [
        'tipo' => $config['tipo'],
        'valor' => $valor,
        'unidade' => $config['unidade'] ?? ''
    ];
}

$topicsToSubscribe = $mqttConfig['topics_list'];
if (empty($topicsToSubscribe)) {
    $topicsToSubscribe[] = $mqttConfig['wildcard_topic'];
}

echo "[" . date('Y-m-d H:i:s') . "] Iniciando cliente MQTT assinante...\n";
echo "[" . date('Y-m-d H:i:s') . "] Broker: {$mqttConfig['broker']}:{$mqttConfig['port']}\n";
echo "[" . date('Y-m-d H:i:s') . "] Tópicos: " . implode(', ', $topicsToSubscribe) . "\n";
echo str_repeat('-', 60) . "\n";

$conn = db_connect();

$cafile = $mqttConfig['tls_enabled'] ? $mqttConfig['tls_cafile'] : null;
$mqtt = new Bluerhinos\phpMQTT($mqttConfig['broker'], $mqttConfig['port'], $mqttConfig['client_id'], $cafile);

$username = $mqttConfig['username'] ?? '';
$password = $mqttConfig['password'] ?? '';

if (!$mqtt->connect(true, null, $username, $password)) {
    die("[ERRO] Não foi possível conectar ao broker MQTT\n");
}

echo "[" . date('Y-m-d H:i:s') . "] ✓ Conectado ao broker com sucesso!\n";

$callback = function ($topic, $msg) use ($conn, $sensorTopicMap) {
    $timestamp = date('Y-m-d H:i:s');
    echo "\n[$timestamp] Nova mensagem recebida!\n";
    echo "  Tópico: $topic\n";
    echo "  Payload: $msg\n";

    try {
        $measurement = extractMeasurement($topic, (string)$msg, $sensorTopicMap);
    } catch (InvalidArgumentException $e) {
        echo "  [ERRO] " . $e->getMessage() . "\n";
        return;
    }

    $tipo = $measurement['tipo'];
    $valor = $measurement['valor'];
    $unidade = $measurement['unidade'] ?: 'valor';

    $stmt = $conn->prepare("SELECT id FROM sensor WHERE tipo = ? AND status = 'ativo' LIMIT 1");
    $stmt->bind_param('s', $tipo);
    $stmt->execute();
    $result = $stmt->get_result();
    $sensor = $result->fetch_assoc();
    $stmt->close();

    if (!$sensor) {
        echo "  [AVISO] Sensor tipo '$tipo' não encontrado - criando novo sensor...\n";

        $descricao = "Sensor $tipo - Auto-criado via MQTT";
        $stmtInsert = $conn->prepare("INSERT INTO sensor (tipo, descricao, status) VALUES (?, ?, 'ativo')");
        $stmtInsert->bind_param('ss', $tipo, $descricao);

        if ($stmtInsert->execute()) {
            $id_sensor = $stmtInsert->insert_id;
            echo "  [OK] Novo sensor criado com ID: $id_sensor\n";
        } else {
            echo "  [ERRO] Falha ao criar sensor: " . $stmtInsert->error . "\n";
            $stmtInsert->close();
            return;
        }
        $stmtInsert->close();
    } else {
        $id_sensor = $sensor['id'];
    }

    $stmtData = $conn->prepare("INSERT INTO sensor_data (id_sensor, valor, unidade, data_hora) VALUES (?, ?, ?, NOW())");
    $stmtData->bind_param('ids', $id_sensor, $valor, $unidade);

    if ($stmtData->execute()) {
        echo "  [✓] Dados salvos no banco com sucesso!\n";
        echo "      ID Sensor: $id_sensor | Valor: $valor $unidade\n";
    } else {
        echo "  [ERRO] Falha ao salvar dados: " . $stmtData->error . "\n";
    }
    $stmtData->close();
};

$subscriptions = [];
foreach ($topicsToSubscribe as $topic) {
    $subscriptions[$topic] = [
        'qos' => 0,
        'function' => $callback
    ];
}

$mqtt->subscribe($subscriptions, 0);

echo "[" . date('Y-m-d H:i:s') . "] ✓ Inscrito nos tópicos configurados!\n";
echo "[" . date('Y-m-d H:i:s') . "] Aguardando mensagens... (Ctrl+C para sair)\n";
echo str_repeat('-', 60) . "\n";

while (true) {
    $mqtt->proc();
    usleep(100000);
}

$mqtt->close();
